Aflores plymouth (#234)
osisoft/Edge-Data-Store-Docs: content/shared-content/shared-content/configuration/egress-endpoints.md

---
uid: EgressEndpointsConfiguration
---

# Egress endpoints

> PI adapters collect time series data, which they can send to a permanent data store (endpoint). This operation is called data egress. The following endpoints are available for data egress:

Reviewer: ```suggestion
PI adapters collect time series data that they can send to a permanent data store (endpoint). This operation is called data egress. The following endpoints are available for data egress:
```

> - AVEVA Data Hub
> - PI servers through PI Web API

Reviewer: ```suggestion
- AVEVA Data Hub

- PI servers through PI Web API
```

For long term storage and analysis, you can configure any adapter to send time series data to one or several of these endpoints in any combination. An egress endpoint is comprised of the properties specified under [Egress endpoint parameters](#egress-endpoint-parameters).

Data egress to a PI server creates a PI point in the PI adapter configuration. Data egress to AVEVA Data Hub creates a stream in the PI adapter configuration.

> The name of the PI point or AVEVA Data Hub stream is a combination of the StreamIdPrefix specified in the adapter data source configuration and the StreamId specified in the adapter data selection configuration.

Reviewer: ```suggestion
The name of the PI point or AVEVA Data Hub stream is a combination of the `StreamIdPrefix` specified in the adapter data source configuration and the `StreamId` specified in the adapter data selection configuration.
```

## Configure egress endpoints

Complete the following steps to configure egress endpoints. Use the `PUT` method in conjunction with the `http://localhost:5590/api/v1/configuration/OmfEgress/dataendpoints` REST endpoint to initialize the configuration.

> 1. Using a text editor, create an empty text file.
> 
> 2. Copy and paste an example configuration for egress endpoints into the file.
> 
>     For sample JSON, see [Examples](#examples).
> 
> 3. Update the example JSON parameters for your environment.
> 
>     For a table of all available parameters, see [Egress endpoint parameters](#egress-endpoint-parameters).
> 
> 4. Save the file. For example, as `ConfigureEgressEndpoints.json`.
> 
> 5. Open a command line session. Change directory to the location of `ConfigureEgressEndpoints.json`.
> 
> 6. Enter the following cURL command (which uses the `PUT` method) to initialize the egress endpoints configuration.

Reviewer: ```suggestion
1. Using a text editor, create an empty text file.

1. Copy and paste an example configuration for egress endpoints into the file.

    For sample JSON, see [Examples](#examples).

1. Update the example JSON parameters for your environment.

    For a table of all available parameters, see [Egress endpoint parameters](#egress-endpoint-parameters).

1. Save the file. For example, as `ConfigureEgressEndpoints.json`.

1. Open a command line session. Change directory to the location of `ConfigureEgressEndpoints.json`.

1. Enter the following `curl` command (which uses the `PUT` method) to initialize the egress endpoints configuration.
```

    ```bash
    curl -d "@ConfigureEgressEndpoints.json" -H "Content-Type: application/json" -X PUT "http://localhost:5590/api/v1/configuration/OmfEgress/dataendpoints"
    ```

>     **Note:**: If you installed the adapter to listen on a non-default port, update `5590` to the port number in use. For a list of other REST operations you can perform, like updating or replacing an egress endpoints configuration, see [REST URLs](#rest-urls).

Reviewer: ```suggestion
    **Note:**: If you installed the adapter to listen on a non-default port, update `5590` to the port number you used. For a list of other REST operations you can perform, like updating or replacing an egress endpoints configuration, see [REST URLs](#rest-urls).
```

    <br/>
    <br/>

## Egress endpoint configuration schema

The full schema definition for the egress endpoint configuration is in the `OmfEgress_DataEndpoints_schema.json` file located in one of the following folders:

> * **Windows**: `%ProgramFiles%\OSIsoft\Adapters\<AdapterName>\Schemas`
> * **Linux**: `/opt/OSIsoft/Adapters/<AdapterName>/Schemas`

Reviewer: ```suggestion
* **Windows**: `%ProgramFiles%\OSIsoft\Adapters\<AdapterName>\Schemas`

* **Linux**: `/opt/OSIsoft/Adapters/<AdapterName>/Schemas`
```

| api/v1/configuration/omfegress/DataEndpoints      | `DELETE`    | Deletes all configured egress endpoints |
| api/v1/configuration/omfegress/DataEndpoints      | `POST`      | Adds an array of egress endpoints or a single endpoint. Fails if any endpoint already exists |
| api/v1/configuration/omfegress/DataEndpoints      | `PUT`       | Replaces all egress endpoints |
| api/v1/configuration/omfegress/DataEndpoints      | `PATCH`     | Allows partial updating of configured endpoints.<br>**Note:** The request must be an array containing one or more endpoints. Each endpoint in the array must include its *Id*. |
| api/v1/configuration/omfegress/DataEndpoints/{Id} | `GET`       | Gets configured endpoint by *Id* |
| api/v1/configuration/omfegress/DataEndpoints/{Id} | `DELETE`    | Deletes configured endpoint by *Id* |
| api/v1/configuration/omfegress/DataEndpoints/{Id} | `PUT`       | Updates or creates a new endpoint with the specified *Id* |
| api/v1/configuration/omfegress/DataEndpoints/{Id} | `PATCH`     | Allows partial updating of configured endpoint by *Id* |

## Egress execution details

> - After configuring an egress endpoint, egress is immediately run for that endpoint. Egress is handled individually per configured endpoint. When data is egressed for the first time, types and containers are egressed to the configured endpoint. After that, only new or changed types or containers are egressed. Type creation must be successful in order to create containers. Container creation must be successful in order to egress data.

Reviewer: ```suggestion
- After you configure an egress endpoint, egress is immediately run for that endpoint. Egress is handled individually per configured endpoint. When data is egressed for the first time, types and containers are egressed to the configured endpoint. After that, only new or changed types or containers are egressed. Type creation must be successful in order to create containers. Container creation must be successful in order to egress data.
```
